src/mono/mono/eventpipe/ds-server.c

```diff
@@ -0,0 +1,317 @@
+#ifdef HOST_WIN32
+#include "ds-ipc-win32.c"
+#else
+#include "ds-ipc-posix.c"
+#endif
+#include "ds-protocol.c"
+#include "ds-eventpipe-protocol.c"
+#include "ds-process-protocol.c"
+#include "ds-dump-protocol.c"
+#include "ds-profiler-protocol.c"
+#else
+#define DS_IMPL_SERVER_GETTER_SETTER
+#include "ds-server.h"
+#include "ds-ipc.h"
+#include "ds-protocol.h"
+#include "ds-process-protocol.h"
+#include "ds-eventpipe-protocol.h"
+#include "ds-dump-protocol.h"
+#include "ds-profiler-protocol.h"
+#include "ep-stream.h"
+#endif
+
+#ifdef FEATURE_AUTO_TRACE
+// TODO: Implement
+#include "ds-autotrace.h"
+#endif
+
+/*
+ * Globals and volatile access functions.
+ */
+
+static volatile uint32_t _server_shutting_down_state = 0;
+static ep_rt_wait_event_handle_t _server_resume_runtime_startup_event = { 0 };
+static bool _server_disabled = false;
+
+static
+inline
+bool
+server_volatile_load_shutting_down_state (void)
+{
+	return (ep_rt_volatile_load_uint32_t (&_server_shutting_down_state) != 0) ? true : false;
+}
+
+static
+inline
+void
+server_volatile_store_shutting_down_state (bool state)
+{
+	ep_rt_volatile_store_uint32_t (&_server_shutting_down_state, state ? 1 : 0);
+}
+
+/*
+ * Forward declares of all static functions.
+ */
+
+static
+void
+server_error_callback_create (
+	const ep_char8_t *message,
+	uint32_t code);
+
+static
+void
+server_error_callback_close (
+	const ep_char8_t *message,
+	uint32_t code);
+
+static
+void
+server_warning_callback (
+	const ep_char8_t *message,
+	uint32_t code);
+
+static
+void
+server_protocol_helper_unknown_command (
+	DiagnosticsIpcMessage *message,
+	DiagnosticsIpcStream *stream);
+
+/*
+ * DiagnosticServer.
+ */
+
+static
+void
+server_error_callback_create (
+	const ep_char8_t *message,
+	uint32_t code)
+{
+	EP_ASSERT (message != NULL);
+	DS_LOG_ERROR_2 ("Failed to create diagnostic IPC: error (%d): %s.\n", code, message);
+}
+
+static
+void
+server_error_callback_close (
+	const ep_char8_t *message,
+	uint32_t code)
+{
+	EP_ASSERT (message != NULL);
+	DS_LOG_ERROR_2 ("Failed to close diagnostic IPC: error (%d): %s.\n", code, message);
+}
+
+static
+void
+server_protocol_helper_unknown_command (
+	DiagnosticsIpcMessage *message,
+	DiagnosticsIpcStream *stream)
+{
+	DS_LOG_WARNING_1 ("Received unknown request type (%d)\n", ds_ipc_message_header_get_commandset (ds_ipc_message_get_header (&message)));
+	ds_ipc_message_send_error (stream, DS_IPC_E_UNKNOWN_COMMAND);
+	ds_ipc_stream_free (stream);
+}
+
+static
+void
+server_warning_callback (
+	const ep_char8_t *message,
+	uint32_t code)
+{
+	EP_ASSERT (message != NULL);
+	DS_LOG_WARNING_2 ("warning (%d): %s.\n", code, message);
+}
+
+EP_RT_DEFINE_THREAD_FUNC (server_thread)
+{
+	EP_ASSERT (server_volatile_load_shutting_down_state () == true || ds_ipc_stream_factory_has_active_ports () == true);
+
+	if (!ds_ipc_stream_factory_has_active_ports ()) {
+		DS_LOG_ERROR_0 ("Diagnostics IPC listener was undefined\n");
+		return 1;
+	}
+
+	ep_rt_thread_setup (true);
```

Comment:
do we need to set some flags on the thread? Something like what the log profiler helper threads get:
```
	mono_thread_info_set_flags (MONO_THREAD_INFO_FLAGS_NO_GC | MONO_THREAD_INFO_FLAGS_NO_SAMPLE);
```

----

 (I'm assuming we call down into the functions that use `DS_ENTER_BLOCKING_PAL_SECTION;` from this thread -  but maybe i am misunderstanding how it's used)

Reply:
I guess, in general, what's the expected interaction between GC STW and the diagnostic server or eventpipe threads?

Reply:
The diagnostic server thread will end up calling through the Mono shim as well as its own PAL for IPC communication. The code is mainly GC safe and code makes no assumptions regarding that, but due to current differences between Mono and CoreCLR in that aspect the EventPipe/DaignosticServer code is setup to at least inform the shim when things occurs that might affect current threads ability to interact with GC in one way or another.

For Mono the diagnostic server thread might end up in EventPipe, Mono runtime as well as it's own IPC PAL code. The thread is therefore setup to be attached in GC unsafe mode at the moment, but will look into running all in GC safe mode in future, but due to dependency on Mono's runtime layer assuming GC safe, it was initial simpler to run none blocking EventPipe/DiagnosticServer code using GC unsafe. Due to that I also believe it needs to be part of STW since it might hit code paths in runtime currently being GC unsafe.

For the IPC streaming threads they will run EventPipe code as well as IPC PAL, but since it might end up in Mono runtime shim as well (using locking, events etc), it is also running in GC unsafe mode and probably need to be part of STW as well since it might hot code paths in runtime currently begin GC unsafe.

In order to make the Mono shim run in GC safe, meaning that we could run EventPipe/DiagnosticServer in GC safe as well means to change several of the Mono runtime methods to be GC safe as well and that is something we could look at going forward.

Excluding these threads from sampling is a good idea, will add that flag for the background threads.